### PR TITLE
Fix `Restart-Computer` to work on localhost when WinRM is not present

### LINKED ISSUE #8197: Restart-Computer does not function on arm64
Steps to reproduce
------------------

Run pwsh on WinPE, ARM64
Run Restart-Computer

X:\Program Files\PowerShell>pwsh
PowerShell 6.1.0
Copyright (c) Microsoft Corporation. All rights reserved.

https://aka.ms/pscore6-docs
Type 'help' to get help.

PS X:\Program Files\PowerShell> Restart-Computer
Restart-Computer : Failed to restart the computer minint-0ehkrmp with the following error message: CimSessionOptions instance is invalid: unsupported protocol or unrecognized custom option.  Verify parameter passed to the CimSessionOptions(string) constr
At line:1 char:1
+ Restart-Computer
+ ~~~~~~~~~~~~~~~~
+ CategoryInfo          : OperationStopped: (minint-0ehkrmp:String) [Restart-Computer], InvalidOperationException
+ FullyQualifiedErrorId : RestartcomputerFailed,Microsoft.PowerShell.Commands.RestartComputerCommand

PS X:\Program Files\PowerShell>

Expected behavior
-----------------

Computer restarts

Actual behavior
---------------

Failure see above

Environment data
----------------

X:\Program Files\PowerShell>pwsh
PowerShell 6.1.0
Copyright (c) Microsoft Corporation. All rights reserved.

https://aka.ms/pscore6-docs
Type 'help' to get help.

PS X:\Program Files\PowerShell> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      6.1.0
PSEdition                      Core
GitCommitId                    6.1.0
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

PS X:\Program Files\PowerShell>

### PR DESCRIPTION
**Pull request by adityapatwardhan**
Fixes #8197 

# PR Summary

PowerShell core uses WSMan transport for all operation. When winRM is not present creating a session fails. The PR uses the default transport when the target computer is localhost. We get the instance on `Win32_ComputerSystem` and then call the instance method on it to restart the computer.

The PR also fixes `Rename-Computer` similarly. 

## PR Context

On winPE arm64, winRM is not present and hence `Restart-Computer` fails.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` to your commit messages if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Comment by adityapatwardhan** [comment]
The change has been manually tested on winPE arm64 machine.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 1]
```diff
@@ -1503,12 +1503,28 @@ private void DoRenameComputerWsman(string computer, string computerName, string
...
-                                ComputerWMIHelper.CimOperatingSystemNamespace,
-                                cimInstance,
-                                "Rename",
-                                methodParameters,
-                                operationOptions);
+                            CimMethodResult result = null;
+
+                            if (isLocalhost)
+                            {
+                                var computerSystem = cimSession.EnumerateInstances(ComputerWMIHelper.CimOperatingSystemNamespace, ComputerWMIHelper.WMI_Class_OperatingSystem);
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Should add a comment that Win32_OperatingSystem is a singleton

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 2]
```diff
@@ -1503,12 +1503,28 @@ private void DoRenameComputerWsman(string computer, string computerName, string
...
+
+                                result = cimSession.InvokeMethod(
+                                    ComputerWMIHelper.CimOperatingSystemNamespace,
+                                    computerSystem.FirstOrDefault(),
+                                    ComputerWMIHelper.CimOperatingSystemShutdownMethod,
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Is this correct?  This is calling Shutdown, but the else case is calling rename?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 3]
```diff
@@ -2133,12 +2149,28 @@ internal static bool SkipSystemRestoreOperationForARMPlatform(PSCmdlet cmdlet)
...
-                            ComputerWMIHelper.CimOperatingSystemNamespace,
-                            ComputerWMIHelper.WMI_Class_OperatingSystem,
-                            ComputerWMIHelper.CimOperatingSystemShutdownMethod,
-                            methodParameters,
-                            operationOptions);
+                        CimMethodResult result = null;
+
+                        if (isLocalhost)
+                        {
+                            var computerSystem = cimSession.EnumerateInstances(ComputerWMIHelper.CimOperatingSystemNamespace, ComputerWMIHelper.WMI_Class_OperatingSystem);
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Add comment that Win32_OperatingSystem is a singleton

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs (hunk 1 above)
Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Remove this piece of code as no change is needed in `Rename-Computer`

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs (hunk 2 above)
Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Remove this piece of code as no change is needed in `Rename-Computer`

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs (hunk 3 above)
Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Added.

**Comment by adityapatwardhan** [comment]
@SteveL-MSFT Please re-review, comments addressed. `Rename-Computer` change reverted as it is not required.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 4]
```diff
@@ -1504,11 +1504,11 @@ private void DoRenameComputerWsman(string computer, string computerName, string
...
-                                ComputerWMIHelper.CimOperatingSystemNamespace,
-                                cimInstance,
-                                "Rename",
-                                methodParameters,
-                                operationOptions);
+                                    ComputerWMIHelper.CimOperatingSystemNamespace,
+                                    cimInstance,
+                                    "Rename",
+                                    methodParameters,
+                                    operationOptions);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Why we change formatting?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs (hunk 3 above)
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: I'd prefer:
```suggestion
                            // Win32_ComputerSystem is a singleton hence FirstOrDefault() return the only instance returned by EnumerateInstances.
                            var computerSystem = cimSession.EnumerateInstances(ComputerWMIHelper.CimOperatingSystemNamespace, ComputerWMIHelper.WMI_Class_OperatingSystem).FirstOrDefault();
```

**Review by iSazonov — COMMENTED** [review]
How would we test this in CIs?

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs (hunk 4 above)
Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: reverted.

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs [hunk 5]
```diff
@@ -2133,12 +2133,30 @@ internal static bool SkipSystemRestoreOperationForARMPlatform(PSCmdlet cmdlet)
...
+
+                        if (isLocalhost)
+                        {
+                            // Win32_ComputerSystem is a singleton hence FirstOrDefault() return the only instance returned by EnumerateInstances.
+                            var computerSystem = cimSession.EnumerateInstances(ComputerWMIHelper.CimOperatingSystemNamespace, ComputerWMIHelper.WMI_Class_OperatingSystem).FirstOrDefault();
+
+                            // Win32_ComputerSystem is a singleton hence FirstOrDefault() return the only instance returned by EnumerateInstances.
+                            result = cimSession.InvokeMethod(
+                                ComputerWMIHelper.CimOperatingSystemNamespace,
+                                computerSystem.FirstOrDefault(),
```
Comment (SteveL-MSFT) [inline src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: The `FirstOrDefault()` call here is redundant

**Review by SteveL-MSFT — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs (hunk 5 above)
Comment (adityapatwardhan) [reply src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs]: Removed duplicate and the duplicate comment as well.

**Review by SteveL-MSFT — APPROVED** [review]

**Review by iSazonov — APPROVED** [review]

**Comment by adityapatwardhan** [comment]
@iSazonov - We cannot test this in CI. This specifically repro on special hardware which is not available in CI system. I have tested it manually.

**Pull request merged**